### ISSUE TITLE
Disappearing viscosity term in the Poisson-pressure equation in the discrete version

### ISSUE THREAD

**Issue by Beramos** [barbagroup/CFDPython#77]
Thank you for this great course! I have been using this course to teach CFD-concepts in Belgium.

I'm stuck with the following question. In video lecture 11, a poisson equation is obtained to correct for the divergence in the velocity. 

![image](https://user-images.githubusercontent.com/11618908/95980638-bf39aa80-0e1d-11eb-87f9-be1cfde17bbc.png)

However in notebook 14 the discretised pressure-Poisson equation looks completely different than the one obtained 
in the video lecture: https://youtu.be/ZjfxA3qq2Lg?t=1647

My question is: what happened in between the curved brackets in the video lecture and equation 13 in the notebook.

**Comment by piyueh** [comment]
Hi @Beramos 
If you were asking about the viscous term in the pressure Poisson equation, then that tern is zero due to the divergence-free condition. 

Theoretically speaking, the conservation of mass in the previous time step <img src="https://render.githubusercontent.com/render/math?math=n"> should already be satisfied, i.e., <img src="https://render.githubusercontent.com/render/math?math=\nabla\cdot\vec{u}^n=0">. So when we apply divergence to the discretized mementum equation shown in this course, the viscous term becomes <img src="https://render.githubusercontent.com/render/math?math=\nu\nabla\cdot\left(\nabla^2\vec{u}^2\right)=\nu\nabla^2\left(\nabla\cdot\vec{u}^n\right)=0">.

**Comment by Beramos** [comment]
I'm not convinced by that statemetn since the whole point is that div(un) should be zero but is not zero due to numerical erros? That why we enforce div(un+1) to be zero through the poisson-pressure equation. 
![image](https://user-images.githubusercontent.com/11618908/96105806-bd81ec80-0eda-11eb-8925-69a70cab1c72.png)

I'm aware I'm being annoying but I'm looking for a sound explanation for my students :angel:

**Comment by labarba** [comment]
Does this post help?
https://lorenabarba.com/blog/cfd-python-12-steps-to-navier-stokes/#comment-1405089241

**Comment by Beramos** [comment]
@labarba, Thanks for taking the time to respond.
I've seen the post in my quest to better understanding the rationale. 

Quoting your response:

> This derivation is the easiest to understand and follow, and that's why we do it in the lesson. But we have to realize that in the *discrete equations* we cannot be sure that the divergence of the velocity is zero, because of possible numerical errors. So we keep the divergence terms —at least some of them— and we don't' eliminate the time derivative term.

Based on which criterium did you decide to keep **some of the divergence terms**

**Comment by labarba** [comment]
The pressure Poisson equation has caused a great deal of arguments in CFD. There are some subtle issues with it.

In the discrete equations we cannot be sure that the divergence of the velocity is zero, because of possible numerical errors. So we keep the divergence terms—at least some of them—and we don't eliminate the time derivative term.

We work with a semi-discretized form of the momentum equation (in vector form) towards the end of Video Lesson 11, and we kept all the terms with divergence, viewing the pressure Poisson equation as a correction.

**Comment by labarba** [comment]
But there are several variations on this... 

Historically, two papers independently introduced a pressure equation to solve incompressible Navier-Stokes: Harlow & Welch (1965), and Chorin (1968).

Chorin proposes a fractional step approach whereby the N-S eq is solved in two steps: (1) ignore the pressure term in the momentum equation, and get a "temporary velocity", u⃗⋆ ; (2) correct this velocity by accounting for the pressure gradient and continuity. This leads to another Poisson equation for pressure that looks like this:
![chorinPPE](https://user-images.githubusercontent.com/4800109/96147691-8bba5700-0ed5-11eb-8f5a-2d282bf6727b.png)

Harlow & Welch work with fully discretized equations, and keep ALL the terms with divergence of velocity. They say that, in principle, the PPE without all the divergence terms could be used, but they add:

> "In practice [the use of all the terms] is desirable in that the solution of [the discrete PPE] need not be nearly so accurately derived to keep the accumulation of compressional discrepancy to a sufficiently low level. Since [the PPE] is solved by an iterative procedure, there is considerable economy in computer usage resulting from any process which decreases the required accuracy of solution. Tests have shown, however, that with a very stringent convergence requirement, the cumulative results of a calculation are independent of whether [all divergence terms are kept or ignored in the PPE]."

Another paper that illustrates the "confusion regarding the use of a Poisson equation for pressure" is Roache (1988). He addresses the important question of boundary conditions, as well as adding that:

> ... a significant advantage of the MAC method [by Harlow & Welch] is that incomplete iterative convergence can be prevented from introducing non-linear instability by the ingenious device of retaining a time derivative for the dilatation, which in the continuum is identically zero, and setting the dilatation to zero at the advanced time step."

In conclusion, we could have neglected all the divergence terms in the equation used for Steps 11 and 12, and you would have been happy. Instead we left one term; we could have left them all, but that would have been a lot more work for you.

**Comment by Beramos** [comment]
Thank you for going to great lengths explaining this. That is a very
satisfying answer.

Op do 15 okt. 2020 17:02 schreef Lorena A. Barba <notifications@github.com>:

**Comment by labarba** [comment]
Out of curiosity, what course do you teach, and how many students is this reaching? We get very little glimpse into hour our public materials are used, and it is satisfying to know about it...

**Comment by piyueh** [comment]
@Beramos Thanks for this question. It motivated me to re-visit the original paper from Harlow and Welch. I agree with @labarba.

> Harlow, F. H., & Welch, J. E. (1965). Numerical calculation of time‐dependent viscous incompressible flow of fluid with free surface. The physics of fluids, 8(12), 2182-2189.

Harlow and Welch suggested keeping all divergence-related terms at time n because it is then unnecessary to solve the pressure accurately. Also, the error accumulation in time caused by nonzero divergence can be contained -- when solving the divergence-free at n+1 (i.e., solving the PPE), we always consider and relieve the divergence error introduced from time n.

However, assume that 1). we solve the pressure accurately, 2). discretized boundary conditions are divergence-free, and 3). initial conditions are also divergence-free. The divergence can then also be assumed to be zero. It should not matter how many divergence-related terms of time step n we drop from the original PPE. Harlow and Welch mentioned this.

We assume the divergence at n is zero, and then drop them. But we actually do not solve the pressure to an extremely accurate state. Then the divergence-free assumption is invalid. This does sound somehow odd. However, for me, I'll say it's just a numerical error. Every numerical scheme has its limitations, assumptions, approximations, and numerical errors. And I feel like the dropped non-free divergence is just one of them.

The step 11 drops only the divergence term in the viscous term but keeps the one in the temporal derivative. I guess the reason is the implementation difficulty due to the boundary nodes. This is an entry-level teaching material, so it tries to keep things simple. And by considering at least one divergence term, it may still give the desired stability property (I think. No proof.).

**Comment by Beramos** [comment]
@piyueh @labarba Thank you for the clarification. The 12 steps to Navier-Stokes course is a module in a practical, advanced modelling course taught at the faculty of bio-science engineering at Ghent University, Belgium. This year there are 20 students taking this course. I've been using your course material since last year and it has been a pleasure using it since it is very self-explanatory. I've added a couple of things, like implicit methods (using the method of lines as an example) but it is mostly your course material
